Add blueprints to wp-now provider

gcsecsey opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "Fixes" keyword and use "Related to" instead.
-->

## Related issues

- Fixes STU-792

## Proposed Changes

- Updated `ServerOptions` and `WordPressServerOptions` interfaces to include blueprint parameter
- Modified `WpNowProvider` to pass blueprint data to wp-now
- Updated `SiteServer` to pass blueprint data from metadata to providers
- Extended `PlaygroundCliProvider` with `PlaygroundCliServerOptions` interface for backwards compatibility

## Testing Instructions
 
- Run `npm start`
- Turn off the `Enable Blueprints` feature flag (this is just turns off Playground CLI provider)

### Creating sites
2. Create a new site and select a blueprint
3. Verify the site creates successfully and the blueprint runs
4. Check that wp-now provider is being used (you don't see `[playground-cli]` prefix in the logs)
5. Test with different blueprints

```
Would have bumped stat: studio-app-site-create=remote-blueprint
Verifying checksums for WordPress version latest...
Starting server for 'My Calm Website'
[31408:0916/115917.975214:ERROR:node_bindings.cc(381)] Most NODE_OPTIONs are not supported in packaged apps. See documentation for more details.
directory: /Users/gcsecsey/Studio/my-calm-website
mode: wordpress
php: 8.3
wp: latest
blueprint steps: 12
Blueprint step completed: login
Blueprint step completed: installPlugin
Blueprint step completed: installPlugin
Blueprint step completed: installPlugin
Blueprint step completed: installPlugin
Blueprint step completed: installPlugin
Blueprint step completed: installPlugin
Blueprint step completed: installPlugin
Blueprint step completed: installPlugin
Blueprint step completed: installPlugin
Blueprint step completed: installTheme
Blueprint step completed: setSiteOptions
Server running at http://localhost:8887
Server started for 'My Calm Website'
```

### Starting/stopping server
1. Stop the server of a site using blueprints
2. Start the server again
3. Check in the logs that the blueprint isn't executed again

```
Starting server for 'My Calm Website'
[31523:0916/115949.788057:ERROR:node_bindings.cc(381)] Most NODE_OPTIONs are not supported in packaged apps. See documentation for more details.
directory: /Users/gcsecsey/Studio/my-calm-website
mode: wordpress
php: 8.3
wp: latest
Server running at http://localhost:8887
Server started for 'My Calm Website'
```

### Existing functionality
1. Create sites without blueprints
4. Verify everything works as before

gcsecsey committed:
use wpnow provider with blueprints

gcsecsey committed:
add feature flag to use wp-now

gcsecsey committed:
revert playground-cli changes

gcsecsey committed:
undo adding new feature flag

gcsecsey committed:
remove comment

wojtekn commented:
@gcsecsey It's great that such a small change allowed to add blueprint support to wp-now flow.

Does the blueprint validation still work? Should we change the default value of the flag to `false` in this PR to let us test the built package, too?

bcotrim reviewed (changes requested):
Thanks for working on this change @gcsecsey 
When I turn off the `Enable Blueprints` flag I don't see the blueprint option in start site.

<img width="2024" height="1424" alt="Image" src="https://github.com/user-attachments/assets/93ad91fc-24ae-4e0d-8eaa-ab0f705b8289" />

Starting a server from a blueprint seems to be working as expected, however I noticed the blueprint is running on every `start`. I think we need to somehow update the server options after we successfully run the blueprints once.

<img width="1030" height="1468" alt="Image" src="https://github.com/user-attachments/assets/0352655e-33c3-4cb4-99d3-0e4db70b3c3d" />

gcsecsey commented:
> Does the blueprint validation still work? Should we change the default value of the flag to false in this PR to let us test the built package, too?

Yes, the validation still works as before:
<img width="2100" height="1424" alt="CleanShot 2025-09-16 at 11 44 55@2x" src="https://github.com/user-attachments/assets/9f7f2f25-8534-43cc-a31d-e478b85c5301" />

Yes, I think setting the flag to false by default is a good idea. I was also thinking maybe renaming the flag because it's not used anymore to control the Blueprints support. 🤔

bcotrim commented:
> Yes, I think setting the flag to false by default is a good idea. I was also thinking maybe renaming the flag because it's not used anymore to control the Blueprints support. 🤔

This sounds good `ENABLE_WP_NOW` or `ENABLE_PLAYGROUND_CLI` seems like a better name, depending on which is going to be released as the default provider.

gcsecsey commented:
> > Yes, I think setting the flag to false by default is a good idea. I was also thinking maybe renaming the flag because it's not used anymore to control the Blueprints support. 🤔
> 
> This sounds good `ENABLE_WP_NOW` or `ENABLE_PLAYGROUND_CLI` seems like a better name, depending on which is going to be released as the default provider.

So as we discussed yesterday, I think we should make wp-now the default for the next release. Based on that, I'd go with `ENABLE_PLAYGROUND_CLI` as the feature flag name.

gcsecsey committed:
always show blueprint option in add site modal

gcsecsey committed:
only run blueprint on site creation

gcsecsey commented:
> Thanks for working on this change @gcsecsey When I turn off the `Enable Blueprints` flag I don't see the blueprint option in start site.
> 
> <img alt="Image" width="2000" height="1424" src="https://private-user-images.githubusercontent.com/8320845/489982577-93ad91fc-24ae-4e0d-8eaa-ab0f705b8289.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTgwMjAxMDksIm5iZiI6MTc1ODAxOTgwOSwicGF0aCI6Ii84MzIwODQ1LzQ4OTk4MjU3Ny05M2FkOTFmYy0yNGFlLTRlMGQtOGVhYS1hYjBmNzA1YjgyODkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDkxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA5MTZUMTA1MDA5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NzhjYTQ0YzZiZGJkYWUzZDNlNzM0NzczYjRkNGRiYzFhNjE2MjYyOGI1NDc4ZGYzM2NlY2U2ZmY1NDZhNDgxMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.zBb2nUeGGm3E9z0WqDlBEe834OWL3lQnju_hxsO9_Rk">
> Starting a server from a blueprint seems to be working as expected, however I noticed the blueprint is running on every `start`. I think we need to somehow update the server options after we successfully run the blueprints once.
> 
> <img alt="Image" width="1030" height="1468" src="https://private-user-images.githubusercontent.com/8320845/489998189-0352655e-33c3-4cb4-99d3-0e4db70b3c3d.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTgwMjAxMDksIm5iZiI6MTc1ODAxOTgwOSwicGF0aCI6Ii84MzIwODQ1LzQ4OTk5ODE4OS0wMzUyNjU1ZS0zM2MzLTRjYjQtOTlkMy0wZTRkYjcwYjNjM2QucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDkxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA5MTZUMTA1MDA5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZGU5ZTkyMjY3NTVkNjhjY2Q2NmZjMzhkZmU1MjFiYjA2ZmNhYmNjNjY2OGU5YWRmMzMxZjA1YTdkMjI2MmY4MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.sgK7sShQ-X5kpzaCF59jf-_FCB3gR-xLNbGAFTDBlxc">

Thanks for highlighting these @bcotrim! I removed the feature flag gate from the blueprints option as both providers should support it now. I also added a new `blueprintExecuted` option to `SiteServerMeta` to guard against running the blueprint on every server restart.

gcsecsey committed:
rename feature flag

bcotrim reviewed (commented):
I've found some more issues

1. The Portfolio blueprint isn't running
2. A simple blueprint with a wp cli command isn't running

[blueprint-wp-cli-image.json](https://github.com/user-attachments/files/22362110/image-test.json)

3. Changing to Playground CLI and running a blueprints fails when trying to start the site. This only happens for new sites created with a blueprint.
```
Error occurred in handler for 'startServer': Error: Could not start server: Error: Invalid blueprint: must NOT have additional properties at
    at /Users/bcotrim/Code/studio/.webpack/main/index.js:329640:40
    at ForkUtilityProcess.<anonymous> (/Users/bcotrim/Code/studio/.webpack/main/index.js:123317:40)
    at ForkUtilityProcess.emit (node:events:530:35)
    at ForkUtilityProcess.a.emit (node:electron/js2c/browser_init:2:73957)
Stopping server with ID 7bb334fd-174a-407b-a528-e7e3761d3c5b
```

wojtekn commented:
> Yes, the validation still works as before:

Thanks. I checked the code and I think that way we violate our provider architecture as wp-now provider uses Playground CLI dependency:

<img width="527" height="202" alt="Screenshot 2025-09-16 at 15 47 52" src="https://github.com/user-attachments/assets/d5a14476-5a3f-4661-84dd-8624abe141f5" />

Should we extract the blueprint validation to provider, and for wp-now have whatever validation it supports?

wojtekn commented:
I spotted one more thing - Studio failed to apply the blueprint, which uses CLI commands. I'm unsure if we can make it work with wp-now, though.:

<img width="526" height="512" alt="Screenshot 2025-09-16 at 15 50 25" src="https://github.com/user-attachments/assets/7356b387-a548-4975-b4fc-e94ad956e0ce" />

gcsecsey commented:
> I've found some more issues
> 
> 1. The Portfolio blueprint isn't running
> 2. A simple blueprint with a wp cli command isn't running
> 
> [blueprint-wp-cli-image.json](https://github.com/user-attachments/files/22362110/image-test.json)
> 
> 3. Changing to Playground CLI and running a blueprints fails when trying to start the site. This only happens for new sites created with a blueprint.
> 
> ```
> Error occurred in handler for 'startServer': Error: Could not start server: Error: Invalid blueprint: must NOT have additional properties at
>     at /Users/bcotrim/Code/studio/.webpack/main/index.js:329640:40
>     at ForkUtilityProcess.<anonymous> (/Users/bcotrim/Code/studio/.webpack/main/index.js:123317:40)
>     at ForkUtilityProcess.emit (node:events:530:35)
>     at ForkUtilityProcess.a.emit (node:electron/js2c/browser_init:2:73957)
> Stopping server with ID 7bb334fd-174a-407b-a528-e7e3761d3c5b
> ```

I've been working on trying to fix 1. for a while, but I didn't get closer to resolving it. This is the error thrown when trying to create a site with the Portfolio blueprint:

```
Error occurred in handler for 'startServer': Error: Error when executing the blueprint step #5 ({"step":"installTheme","themeZipFile":{"resource":"wordpress.org/themes","slug":"neve"},"options":{"activate":true},"themeData":{"resource":"wordpress.org/themes","slug":"neve"}}) : Theme neve could not be activated – WordPress exited with no error. Sometimes, when $_SERVER or site options are not configured correctly, WordPress exits early with a 301 redirect. Inspect the "debug" logs in the console for more details
    at Object.run (/Users/gcsecsey/a8c-projects/studio/.webpack/main/siteServerProcess.js:541438:149811)
    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
    at async zt (/Users/gcsecsey/a8c-projects/studio/.webpack/main/siteServerProcess.js:541438:151558) {
  [cause]: Error: Theme neve could not be activated – WordPress exited with no error. Sometimes, when $_SERVER or site options are not configured correctly, WordPress exits early with a 301 redirect. Inspect the "debug" logs in the console for more details
      at cr (/Users/gcsecsey/a8c-projects/studio/.webpack/main/siteServerProcess.js:541035:109)
      at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
      at async Object.Zr [as installTheme] (/Users/gcsecsey/a8c-projects/studio/.webpack/main/siteServerProcess.js:541315:3361)
      at async y (/Users/gcsecsey/a8c-projects/studio/.webpack/main/siteServerProcess.js:541438:151080)
      at async Object.run (/Users/gcsecsey/a8c-projects/studio/.webpack/main/siteServerProcess.js:541438:149778)
      at async zt (/Users/gcsecsey/a8c-projects/studio/.webpack/main/siteServerProcess.js:541438:151558)
}
```

From this I thought that the [redirection mu plugin](https://github.com/Automattic/studio/blob/4136650226add4da4a295bdf5c5132ba68f6a2ba/vendor/wp-now/src/wp-now.ts#L431) might not work correctly in all cases. However, trying to update this or even removing it didn't affect the error being thrown. @bcotrim do you maybe have some idea what else could be the route cause here? Thanks!

gcsecsey committed:
run wp cli commands in a cli runtime

sejas reviewed (commented):
I was taking a look if I could help with some blocker, and found @gcsecsey  already fixed the wp-cli blueprint step. 😍 

<img width="1196" height="858" alt="Screenshot 2025-09-17 at 12 07 51" src="https://github.com/user-attachments/assets/f76aa729-e42d-4356-9bb2-6aca1c28b36c" />
